networkmanager: Add WiFi interface handler

PedroJG opened the pull request:
Integrated @artur-rs's development as mentioned in #13551 
This is one of the patches separated. This development is NOT mine, and I'm just rebasing it to master, though I would like to keep working on it and expanding it further with you guys' help.

Patch changes:
- added Wifi settings dialog
- integrated "802-11-wireless", "802-1x" and
"802-11-wireless-security" settings from DBus API
- implemented security options (None, WPA-PSK, WPA-EAP TLS
and WPA-EAP PEAP)

PedroJG commented:
It's cleaned up. Should be ready to review now

marusak commented on pkg/networkmanager/index.html:
```diff
@@ -108,56 +108,92 @@ <h2 class="panel-title">
-            <button translate="yes" id="networking-add-team" class="btn btn-default network-privileged">Add Team</button>
-            <button translate="yes" id="networking-add-bridge" class="btn btn-default network-privileged">Add Bridge</button>
-            <button translate="yes" id="networking-add-vlan" class="btn btn-default network-privileged">Add VLAN</button>
+    </div>
```
Seems this end tag is here harmful. Due to it all the following panels are not under `container-fluid` class and it breaks the page.

PedroJG replied:
I have fixed this in the latest commits. There were a cople closing div tags out of place

marusak commented on pkg/networkmanager/interfaces.js:
```diff
@@ -2576,9 +2856,23 @@ PageNetworkInterface.prototype = {
+            return $('<tr>').append($('<td class="network-interface-separator" colspan="100%">').text(title));
+        }
+
+        /* function render_interface_type_row() {
```
cleanup

marusak commented on pkg/networkmanager/interfaces.js:
```diff
@@ -2576,9 +2856,23 @@ PageNetworkInterface.prototype = {
         $('#network-interface-delete').prop('hidden', !is_deletable || !managed);
 
+        function render_interface_section_separator(title) {
+            return $('<tr>').append($('<td class="network-interface-separator" colspan="100%">').text(title));
```
Can you please use Web API instead of jQuery? And please use class and css rather than `colspan` directly here.

PedroJG replied:
Why don't we want to use jQuery here? I see it being used in a few more places along the document, is it a problem with readability?

marusak replied:
We want to reduce amount of jquery we use, so ideally we don't want to write a new code in it.

PedroJG replied:
Okay, I've been able to take jquery out of the equation but since this section is formatted as a table, there is no way for me to take the "colspan" to CSSsince colspan is a structural element in HTML and not a style element and there is no implementation for this in CSS.
Would you rather have the colspan stay, or format the page with divs and flex instead? Or is there another way that I'm not seeing? Thanks!

marusak replied:
When you use web API, then `setAttribute` should be usable for setting up colspan.
 https://developer.mozilla.org/en-US/docs/Web/API/Element/setAttribute

marusak commented on pkg/networkmanager/interfaces.js:
```diff
@@ -2729,6 +3021,35 @@ PageNetworkInterface.prototype = {
                                 .syn_click(self.model, function () { configure() })));
             }
 
+            function render_general_settings_row() {
```
In this function there are some strings that need to be translated (wrap them in `_("string")`.

marusak commented on pkg/networkmanager/interfaces.js:
```diff
@@ -2902,8 +3223,30 @@ PageNetworkInterface.prototype = {
+                if (!options)
+                    return null;
+
+                rows.push(_("SSID: " + (options.ssid ? atob(options.ssid) : "Not defined")));
```
This translation would not work. Please do something like:
`cockpit.format(_("SSID: $0"), options.ssid ? atob(options.ssid) : _("Not defined"))`

PedroJG commented:
![image](https://user-images.githubusercontent.com/50325076/77056678-0fa97080-69d3-11ea-9e40-146bbc4d6d46.png)
This is how "Networking" is looking now

![image](https://user-images.githubusercontent.com/50325076/77056769-3798d400-69d3-11ea-80ab-8da7e8e1021b.png)
This is the "Add wifi" screen without a security option selected

![image](https://user-images.githubusercontent.com/50325076/77056850-5b5c1a00-69d3-11ea-9d3e-94c48d2849ac.png)
"Add wifi" screen with WPA2 Personal

![image](https://user-images.githubusercontent.com/50325076/77056932-76c72500-69d3-11ea-8a22-c4a7f1447de6.png)
"Add wifi" screen with WPA Enterprise

marusak commented:
Thanks for the screenshots! Can you also please post wifi details screenshot?

PedroJG commented:
![image](https://user-images.githubusercontent.com/50325076/77156698-2e296d80-6aa0-11ea-854c-57aa868d0073.png)
This would be the wifi details screen

marusak commented:
Thanks @PedroJG !
@andreasn WDYT about this? Is this usable or you have some ideas what is missing/is not on the right place/is excessive? I would like to firstly know it is going the right way before we start polishing it up.

andreasn commented:
I think it makes sense, yes!

marusak commented on pkg/networkmanager/interfaces.js:
```diff
@@ -2567,9 +2848,16 @@ PageNetworkInterface.prototype = {
         $('#network-interface-delete').prop('hidden', !is_deletable || !managed);
 
+        function render_interface_section_separator(title) {
+            return `<tr>
```
Does this work? I would expect that you would use things like `document.createElement("tr")`...
doc: https://developer.mozilla.org/en-US/docs/Web/API/Document/createElement

PedroJG replied:
There are two ways to make this work. One is using jquery somewhere else (Which I was using at the time of that commit), and another one is to use innerHTML as I do use now

marusak commented on pkg/networkmanager/interfaces.js:
```diff
@@ -3165,6 +3498,150 @@ function connection_devices(con) {
+        }
+
+        function vpn_connections_handler() {
+            var process = cockpit.spawn(["nmcli", "-g", "name,type,uuid", "con"]);
```
I see you already figured it out - `Eliminated nmcli calls for D-Bus alternatives`. Or you still have some question?

marusak commented:
I've played with this a bit and here are some comments:

1. Can you please remove those gray boxes? If you think that theose groups should be there, then please at least format it better. But this seems like out of scope of this PR as the same info is there now, but no such groups are created there.
![removeroupshraders](https://user-images.githubusercontent.com/12330670/77903530-5163d200-7283-11ea-9e3c-83a99ec05b6f.png)

2.  I am wondering about the order of the 'Add new' dropdown. Seems a bit random. Alphabetically would make more sense.

3. In the overview the first few columns are too crowded and the remaining ones have too much space, even though they contain only short information

4. Adn the big one - 'Browse...' for selecting certificates and keys is assuming machine, from which we opened browser. But the main purpose of Cockpit is to be used with remote machine, so machine with browser != machine which you manage. Is the main usecase you want to use the machine from which you browser or the machine on which you set the wifi to contain these files? (yeah yeah, server and wifi... I know...)

PedroJG commented:
> 4. Adn the big one - 'Browse...' for selecting certificates and keys is assuming machine, from which we opened browser. But the main purpose of Cockpit is to be used with remote machine, so machine with browser != machine which you manage. Is the main usecase you want to use the machine from which you browser or the machine on which you set the wifi to contain these files? (yeah yeah, server and wifi... I know...)

It's not really clear to me what you are asking on this comment. I think it's not very intuitive to have a browse button that doesn't actually browse the machine in which you are accessing the website

Also, our usecase is to manage an IoT device with cockpit, which is not really considered a server per se. That's a case in which you would want to have the ability to manage wifi connections and import  certificates from remote locations in my opinion

andreasn commented:
> That's a case in which you would want to have the ability to manage wifi connections and import certificates from remote locations in my opinion

Sounds fair to me. We can always add the ability to select a cert that resides on the server at a later point. I feel part of that would also hinge on the work Simon is doing for general handling of certificates.

marusak commented:
> Also, our usecase is to manage an IoT device with cockpit, which is not really considered a server per se. That's a case in which you would want to have the ability to manage wifi connections and import certificates from remote locations in my opinion

I see.
@andreasn want to check with you - does this use case make sense for cockpit? Is that a correct use of UI elements?

andreasn commented:
Yes, I think it does.
Just to verify, will all wifi-stuff be hidden if there is no wifi devices on the system?

PedroJG commented:
> Just to verify, will all wifi-stuff be hidden if there is no wifi devices on the system?

As of right now you'd see the "Add wifi" option in the dropdown, but is not very disruptive otherwise. This could be hidden in case there was no wifi though

PedroJG commented on pkg/networkmanager/interfaces.js:
```diff
@@ -1050,10 +1147,12 @@ function NetworkManagerModel() {
             "org.freedesktop.NetworkManager.Device.Bridge",
-            "org.freedesktop.NetworkManager.Device.Vlan"
+            "org.freedesktop.NetworkManager.Device.Vlan",
+            "org.freedesktop.NetworkManager.Device.Vpn"
```
It is indeed, my bad!

PedroJG commented on pkg/networkmanager/interfaces.js:
```diff
@@ -1460,7 +1561,7 @@ function render_active_connection(dev, with_link, hide_link_local) {
     }
 
-    return $('<span>').text(parts.join(", "));
+    return `<span> ${parts.join(", ")} </span>`;
```
No, this is actually not jquery. It's JS's template strings
https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Template_literals

PedroJG commented on pkg/networkmanager/interfaces.js:
```diff
@@ -2945,13 +3276,27 @@ PageNetworkInterface.prototype = {
+
+        const information_rows = render_general_information_rows();
+
+        for (var i in information_rows) {
```
That one slip through, fixing it

PedroJG commented on pkg/networkmanager/interfaces.js:
```diff
@@ -3165,6 +3510,134 @@ function connection_devices(con) {
+        const model = PageNetworkGeneralSettings.model;
+        const setts = model.get_settings().Connections;
+
+        function vpn_connections_handler() {
```
This is indeed related to selecting a connection's secondary connection (A VPN from the system).
Now I see this would be better off in a different PR since it's not strictly related to the wifi manager but rather to  networkmanager as a whole. Is it worth splitting?

marusak replied:
I think so. Or at least separate commit in this PR.
(and yes, now here are 13 commits, but should be two (or 3 with this vpn))

PedroJG commented:
@marusak please let me know if there is any new comment. I'm currently working on the testing of this

marusak commented:
@mvollmer you think you could take a quick look at this and give review? Thanks!

PedroJG commented:
I've removed everything not strictly related to wifi handlers and added SSID selector so that you don't have to type in the SSID by hand anymore.

I understand you are probably very busy and this PR is not a high PRIO but unfortunately my time slot here is coming to an end as my company needs me to jump to some other projects :(.

It would be a pity if all this time we both spent on this gets lost so it would be great if you could take a look and get this running upstream as soon as possible.

Thank you beforehand

marusak commented:
```
 pip install pycodestyle
$ pycodestyle ...

  '\n\n'
PASS: tools/test-static-code 7 pep8 test
ERROR: tools/test-static-code process failed: 1
``` 
from [semaphore](https://semaphoreci.com/cockpit-project/cockpit/branches/pull-request-13616/builds/12)
You should be able to reproduce this by running ` ./tools/test-static-code` locally.

mvollmer commented on pkg/networkmanager/index.html:
```diff
@@ -391,6 +392,986 @@ <h2 class="panel-title" translate="yes">Unmanaged Interfaces</h2>
+    </div>
+  </div>
+
+</body><!DOCTYPE html>
```
What's going on here?  index.html has been duplicated somehow.

PedroJG replied:
Sorry, just fixed it. Stupid mistake

mvollmer commented on pkg/networkmanager/interfaces.js:
```diff
@@ -2567,26 +2837,29 @@ PageNetworkInterface.prototype = {
-                        dev.Carrier
-                            ? (dev.Speed ? cockpit.format_bits_per_sec(dev.Speed * 1e6) : _("Yes"))
-                            : _("No")));
+                return `<tr>
```
This change seems unmotivated to me.  IMO it's fine to keep using jQuery here since the rest of the code does it as well. There is no point in introducing two styles.

PedroJG commented:
Is this still in progress?

andreasn commented:
Tried to build this today, but got this build error. Could totally be some shit on my side, though.
https://pastebin.com/6tQAwa5v

PedroJG commented:
> Tried to build this today, but got this build error. Could totally be some shit on my side, though.
> https://pastebin.com/6tQAwa5v

Looks like it LITERALLY is missing a newline, let me add one and you can check again

andreasn commented:
Tested this briefly. On one system with wifi and one without. I get the add wifi on both of them, but only one of them actually had a wifi interface. Is there any way of hiding the add wifi button if there are no wifi interfaces present?
Maybe one way could be to show the wifi interface in the list, and then only be able to set the SSID and the other stuff from there, rather than having device as the dropdown in the add wifi dialog.
Does that make sense as a mental model?

mvollmer commented:
> Maybe one way could be to show the wifi interface in the list, and then only be able to set the SSID and the other stuff from there, rather than having device as the dropdown in the add wifi dialog.
> Does that make sense as a mental model?

I was thinking we could have the "Add Wifi" button on the details page wifi interfaces. That way it's out of the way of people that don't have Wifi interfaces, or are not interested in them. Also, we could call that button "Connect", no?

PedroJG commented:
> Maybe one way could be to show the wifi interface in the list, and then only be able to set the SSID and the other stuff from there, rather than having device as the dropdown in the add wifi dialog.
> Does that make sense as a mental model?

Yeah, that makes a lot of sense, I'll see what I can do and get back to you

PedroJG commented:
@mvollmer @andreasn 


>  I was thinking we could have the "Add Wifi" button on the details page wifi interfaces. That way it's out of the way of people that don't have Wifi interfaces, or are not interested in them. Also, we could call that button "Connect", no?

I did exactly that, take a quick peek and see how it's looking!

andreasn commented:
> I did exactly that, take a quick peek and see how it's looking!

Love it!

I seem to have some issues when I run this with multiple wifi-cards. I have a box of wifi dongles (accumulated over time), so I plugged two of them so I could play around with this patch on my host without disconnecting all the time.
For some reason the disconnected devices think they are attached to the same network as the original, connected one, and when I hit the "Connect to wifi" button, I just get a blank dialog.
![Screenshot from 2020-05-08 12-26-08](https://user-images.githubusercontent.com/209895/81398187-031adc00-9129-11ea-810e-61e31c326893.png)
![Screenshot from 2020-05-08 12-28-09](https://user-images.githubusercontent.com/209895/81398190-044c0900-9129-11ea-96d8-c3935b5d2a0b.png)

PedroJG commented:
@andreasn Should be good now. When facing blank screens it probably means that you're not giving it enough time to load properly. Make sure to give it a couple extra seconds before trying to connect.
I've tested this with up to 3 wifi adapters and gave me good results. Let me know how it's going for you!

PedroJG commented:
@mvollmer @andreasn any updates?

andreasn commented:
Unfortunately I'm getting the same results, still. Blank connection setting dialog, regardless of how long I wait.
I'm also getting some messages about _"Not authorized to control networking."_, even though my user is in the wheel group.

PedroJG commented:
> Unfortunately I'm getting the same results, still. Blank connection setting dialog, regardless of how long I wait.
> I'm also getting some messages about _"Not authorized to control networking."_, even though my user is in the wheel group.

@andreasn Are you able to provide a couple screenshots of the process and errors? I'm not being able to get the same errors you do with various types of users and permissions

PedroJG commented:
This is the case on my end: You either have permissions to control networking, and you get the desired button to connect to a wifi network, like this:
![imagen](https://user-images.githubusercontent.com/50325076/86891270-d8d0a500-c0fe-11ea-8cbc-fedc0cae8568.png)


Or you don't have and don't get the button at all, this way:

![imagen](https://user-images.githubusercontent.com/50325076/86891450-28af6c00-c0ff-11ea-80fc-b4b089003c01.png)


When this is the case, you also don't get to add other types of networking, as is intended:

![imagen](https://user-images.githubusercontent.com/50325076/86891524-4381e080-c0ff-11ea-9ec1-dabb7f5b08bf.png)

> 
> Unfortunately I'm getting the same results, still. Blank connection setting dialog, regardless of how long I wait.
> I'm also getting some messages about "Not authorized to control networking.", even though my user is in the wheel group.

@andreasn Are you able to replicate this so I can take a look and potentially fix it?
Thank you beforehand

andreasn commented:
I'm starting to suspect this needs a rebase against master. The new sudo work @mvollmer did doesn't quite bite.
I have things set up to work on my local machine, since I've been unsuccessful in getting wifi working in any test VMs:
https://github.com/cockpit-project/cockpit/blob/master/HACKING.md#working-on-your-local-machine-cockpits-session-pages

PedroJG commented:
Is there any chance you would test this before it needs another rebase? if not, we could sync on when is possible for you so I can rebase it on that specific moment to avoid leaving it hanging again and require another rebase

andreasn commented:
I was able to get it working now, with some help from @mvollmer!

No connect to wifi buttons though, will keep poking.
![Screenshot from 2020-07-14 13-59-40](https://user-images.githubusercontent.com/209895/87423252-4a669280-c5da-11ea-98a8-a4259866871c.png)

andreasn commented:
I do see this stuff in the DOM though.
```
<div class="panel-actions">
          <button class="pf-c-button pf-m-danger network-privileged" id="network-interface-delete" style="display: inline-block;" hidden="">Delete</button>
<span id="network-interface-delete-switch" class="network-privileged"><label class="onoff-ct"><input type="checkbox" checked=""><span class="switch-toggle"></span></label></span>
</div>
```

But only the ON/OFF button is visible.

andreasn commented:
Note that I do have administrative access. My user account has the System Administrator role (wheel)

PedroJG commented:
> Note that I do have administrative access. My user account has the System Administrator role (wheel)

This is the thing that gets me. I get the whole display of buttons when logging in with administrative access, and I get the same output you do when logging in without them

martinpitt commented:
I built this branch now, and I get the same effect as @andreasn -- the UI is read only, as this branch is too old to work with current cockpit. Indeed it would be easier to rebase to get over the novnc build failure, the authentication rework, and the image-prepare failure.

Cherry-picking commit acd4f87ed74b and building a VM (`test/image-prepare -q fedora-32`) works better. I see no difference without wifi devices. In the VM, a wifi can be simulated with [mac80211_hwsim](https://people.debian.org/~mpitt/talk-simulate-hardware/wifi.sh), but unfortunately that module is not shipped any more in Fedora.

With a Debian VM I can test it with
```
sudo apt install -y hostapd linux-image-amd64
reboot
curl -O https://people.debian.org/~mpitt/talk-simulate-hardware/wifi.sh
chmod a+x wifi.sh
./wifi.sh
```

With that, when I click on "Connect to wifi" button, or the "SSID: Not defined" link, I get a weird dialog:

![image](https://user-images.githubusercontent.com/200109/87916000-28a25b00-ca73-11ea-931c-e5fef787b880.png)

is that still work in progress,  or what is that supposed to do? I suspect it should show me a list of available SSIDs? The "Apply" button does not do anything.

NM sees the simulated wifi from mac80211_hswim and hostapd:
```
root@debian:~# nmcli d wifi list
IN-USE  BSSID              SSID      MODE   CHAN  RATE       SIGNAL  BARS  SECURITY 
        02:00:00:00:00:00  ZE WIFI!  Infra  1     54 Mbit/s  100     ▂▄▆█  WPA1     
```
and on the command line (or nm-applet) I can connect to that just fine:

    nmcli d wifi connect "ZE WIFI!" password 12345678

After that the details update for the SSID, WPA, and such:

![image](https://user-images.githubusercontent.com/200109/87916409-b1b99200-ca73-11ea-90fb-3c8df7c039b3.png)

But the "Connect to wifi" dialog still looks the same.

marusak commented:
Networking page went through a rather big rewrite a few days ago. I am afraid this won't play along with the new code anymore.
Also it is stale and needs a lot of work. This is from series of PRs ( #13616, #13621, #13622) where all are based on some older PR. All of them are based on commit 'Separated general settings' so I would say if someone is still willing to do this, let's do it like this:
1. Open **one** PR with **just this general** commit. (One simpler PR makes it possible for us to review it code-wise, design-wise and test-wise. Having 3 massive PRs all of which are interlocked and bring a lot of functionality makes it super difficult to get through).
2. Once that PR is merged, then open again one PR from those 3. Again, having one PR that brings just some additional functionality makes it much easier to get through.
3.  Once that PR is merged, then open again one PR from those... you get the picture :)
